linera-base/src/crypto/mod.rs

```diff
@@ -46,87 +78,183 @@ pub type ValidatorKeypair = secp256k1::Secp256k1KeyPair;
+    Ed25519(ed25519::Ed25519SecretKey),
+    /// secp256k1 secret key.
+    Secp256k1(secp256k1::Secp256k1SecretKey),
+}
 
 /// The signature of a chain owner.
 #[derive(Eq, PartialEq, Copy, Clone, Debug, Serialize, Deserialize)]
-pub struct AccountSignature(ed25519::Ed25519Signature);
+pub enum AccountSignature {
+    /// Ed25519 signature.
+    Ed25519(ed25519::Ed25519Signature),
+    /// secp256k1 signature.
+    Secp256k1(secp256k1::Secp256k1Signature),
+}
 
 impl AccountSecretKey {
-    #[cfg(all(with_getrandom, with_testing))]
-    /// Generates a new `AccountSecretKey`.
-    pub fn generate() -> Self {
-        Self(ed25519::Ed25519SecretKey::generate())
-    }
-
-    #[cfg(with_getrandom)]
-    /// Returns a new `AccountSecretKey` generated from the given `seed`.
-    pub fn generate_from<R: CryptoRng>(rng: &mut R) -> Self {
-        let secret_key = ed25519::Ed25519SecretKey::generate_from(rng);
-        AccountSecretKey(secret_key)
-    }
-
     /// Returns the public key corresponding to this secret key.
     pub fn public(&self) -> AccountPublicKey {
-        AccountPublicKey(self.0.public())
+        match self {
+            AccountSecretKey::Ed25519(secret) => AccountPublicKey::Ed25519(secret.public()),
+            AccountSecretKey::Secp256k1(secret) => AccountPublicKey::Secp256k1(secret.public()),
+        }
     }
 
     /// Copies the secret key.
     pub fn copy(&self) -> Self {
-        AccountSecretKey(self.0.copy())
+        match self {
+            AccountSecretKey::Ed25519(secret) => AccountSecretKey::Ed25519(secret.copy()),
+            AccountSecretKey::Secp256k1(secret) => AccountSecretKey::Secp256k1(secret.copy()),
+        }
+    }
+
+    /// Creates a signature for the `value` using provided `secret`.
+    pub fn sign<'de, T>(&self, value: &T) -> AccountSignature
+    where
+        T: BcsSignable<'de>,
+    {
+        match self {
+            AccountSecretKey::Ed25519(secret) => {
+                let signature = Ed25519Signature::new(value, secret);
+                AccountSignature::Ed25519(signature)
+            }
+            AccountSecretKey::Secp256k1(secret) => {
+                let signature = secp256k1::Secp256k1Signature::new(value, secret);
+                AccountSignature::Secp256k1(signature)
+            }
+        }
     }
 }
 
 impl AccountPublicKey {
-    #[cfg(with_testing)]
-    /// Constructs a test key from a seed.
-    pub fn test_key(seed: u8) -> Self {
-        Self(ed25519::Ed25519PublicKey::test_key(seed))
+    /// Returns the signature scheme of the public key.
+    pub fn scheme(&self) -> SignatureScheme {
+        match self {
+            AccountPublicKey::Ed25519(_) => SignatureScheme::Ed25519,
+            AccountPublicKey::Secp256k1(_) => SignatureScheme::Secp256k1,
+        }
     }
 
     /// Returns the byte representation of the public key.
-    pub fn to_bytes(&self) -> Vec<u8> {
-        self.0.as_bytes()
+    pub fn as_bytes(&self) -> Vec<u8> {
```

Comment:
Why are re-defining a serialization function here? (also it turns out the bytes are exactly the BCS bytes)

Reply:
Are we re-defining? I'd rather see serialization code call into this. I don't like leaving something as important as serialization to serde. We don't really control what's going on inside and might even not notice when it changes (between versions for example).

Reply:
Done in [Use bcs instead of manually crafting the (de)serialization](https://github.com/linera-io/linera-protocol/pull/3465/commits/f7d33d04b25150fac5fcb5e45b5d914366c13aaa).